### PR TITLE
[Feature] Support standard OpenAI API sampling parameters for thinker stage. (max_tokens...)

### PR DESCRIPTION
**Pull request by wonjerry**
This enables max_tokens, temperature, top_p, top_k, min_p, and other standard OpenAI API parameters to work for the thinker stage.

Fixes: #445

PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Previously, `max_tokens`, `temperature`, `top_p`, and other standard OpenAI API parameters were ignored. Users had to use `sampling_params_list` with parameters for all 3 stages, which was cumbersome.

This PR enables standard OpenAI API parameters to work for the thinker (text generation) stage while maintaining default parameters for other stages (talker, code2wav).

**Behavior:**
- If standard OpenAI params (`max_tokens`, `temperature`, etc.) are provided → applied to thinker stage
- If not provided → uses default values from stage config YAML (`default_sampling_params`)
- Other stages (talker, code2wav) always use YAML defaults

**Changes:**
- Add `_get_thinker_stage_index()` to find the comprehension stage using `is_comprehension` field
- Add `_sampling_params_to_dict()` for SamplingParams to dict conversion
- Add `_build_sampling_params_list_from_request()` to build params list from standard OpenAI parameters
- Use `request.to_sampling_params()` for thinker stage when `sampling_params_list` is not provided

## Test Plan

```bash
  # Test standard OpenAI params (NEW - should limit to ~50 tokens)
  curl http://localhost:8000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
      "model": "Qwen3-Omni-30B-A3B-Instruct",
      "messages": [{"role": "user", "content": "Hello, tell me a long story"}],
      "modalities": ["text"],
      "max_tokens": 50
    }'

  # Test existing sampling_params_list method still works
  curl http://localhost:8000/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
      "model": "Qwen3-Omni-30B-A3B-Instruct",
      "messages": [{"role": "user", "content": "Hello, tell me a long story"}],
      "modalities": ["text"],
      "sampling_params_list": [
        {"max_tokens": 30, "temperature": 0.4},
        {"max_tokens": 4096, "temperature": 0.9},
        {"max_tokens": 65536, "temperature": 0.0}
      ]
    }'

## Test Result
Before: max_tokens parameter ignored, always used default (2048 tokens)

After: max_tokens: 50 respected
{"usage": {"prompt_tokens": 15, "total_tokens": 65, "completion_tokens": 50}}
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 1]
```diff
@@ -407,6 +410,58 @@ def _to_sampling_params_list(self, sampling_params_list: list[dict]) -> list[Sam
                 raise ValueError(f"Invalid sampling params: {sampling_params}")
         return final_sampling_params_list
 
+    def _get_thinker_stage_index(self) -> int:
```
Comment (tzhouam) [inline vllm_omni/entrypoints/openai/serving_chat.py]: the serving chat should work for all the models, could you change it to some name "_get_first_comprehensive_stage_index"

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 2]
```diff
@@ -407,6 +410,58 @@ def _to_sampling_params_list(self, sampling_params_list: list[dict]) -> list[Sam
...
+        for idx, stage in enumerate(self.engine_client.stage_list):
+            if stage.is_comprehension:
+                return idx
+        raise ValueError("No thinker stage (is_comprehension=True) found in stage_list")
```
Comment (tzhouam) [inline vllm_omni/entrypoints/openai/serving_chat.py]: Same as  above

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 3]
```diff
@@ -407,6 +410,56 @@ def _to_sampling_params_list(self, sampling_params_list: list[dict]) -> list[Sam
...
+        import msgspec.structs
+        return msgspec.structs.asdict(params)
+
+    def _build_sampling_params_list_from_request(
+        self,
+        request: ChatCompletionRequest,
+    ) -> list[SamplingParams]:
+        """Build sampling_params_list using standard OpenAI API parameters.
+
+        For the thinker stage, converts OpenAI API parameters (max_tokens, temperature,
```
Comment (tzhouam) [inline vllm_omni/entrypoints/openai/serving_chat.py]: same as above

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 4]
```diff
@@ -407,6 +410,56 @@ def _to_sampling_params_list(self, sampling_params_list: list[dict]) -> list[Sam
...
+        top_p, etc.) to SamplingParams. For other stages, uses default sampling params.
+
+        Args:
+            request: The chat completion request containing OpenAI API parameters.
+
+        Returns:
+            List of SamplingParams, one for each stage.
+        """
+        default_params_list = self.engine_client.default_sampling_params_list
+        thinker_idx = self._get_thinker_stage_index()
```
Comment (tzhouam) [inline vllm_omni/entrypoints/openai/serving_chat.py]: Same name problem

**Comment by tzhouam** [comment]
This looks great! One more small suggestion, besides the naming problem: could you add some tests for parameters other than max_tokens (e.g. temperature)? The code seems correct from reading it, but a few extra tests would make it even more robust.
Thanks a lot for your effort and contribution.

File: vllm_omni/entrypoints/openai/serving_chat.py [hunk 5]
```diff
@@ -407,6 +410,56 @@ def _to_sampling_params_list(self, sampling_params_list: list[dict]) -> list[Sam
...
+        for idx, default_params in enumerate(default_params_list):
+            if idx == thinker_idx:
+                default_dict = self._sampling_params_to_dict(default_params)
+                max_tokens = request.max_tokens or default_params.max_tokens
+                params = request.to_sampling_params(
+                    max_tokens=max_tokens,
+                    logits_processor_pattern=None,
+                    default_sampling_params=default_dict,
+                )
+                sampling_params_list.append(params)
```
Comment (wonjerry) [inline vllm_omni/entrypoints/openai/serving_chat.py]: Could you review this approach and let me know if you have any concerns?


## Problem

The current implementation of `_sampling_params_to_dict()` only includes a subset of SamplingParams fields:

  ```python
  def _sampling_params_to_dict(self, params: SamplingParams) -> dict:
      return {
          "temperature": params.temperature,
          "top_p": params.top_p,
          "top_k": params.top_k,
          "min_p": params.min_p,
          "repetition_penalty": params.repetition_penalty,
          "stop_token_ids": params.stop_token_ids,
          "max_tokens": params.max_tokens,
      }
```

  This dict is passed to request.to_sampling_params(default_sampling_params=...), but to_sampling_params() only uses these fields from default_sampling_params:
  - temperature, top_p, top_k, min_p, repetition_penalty

  Other fields like seed are taken directly from self.seed (the request), not from default_sampling_params. So if a user doesn't specify seed in the request, the final SamplingParams gets seed=None instead of the YAML default (seed=42).

  Result: Stage config YAML defaults (e.g., seed, detokenize) are silently lost when using standard OpenAI API parameters.

## Solution
So, Instead of using request.to_sampling_params(), I:

  1. Clone the default SamplingParams from YAML
  2. Override only the fields that the user explicitly provided in the request (ChatCompletionRequest class)
```python
  def _build_sampling_params_list_from_request(self, request):
      ...
      if idx == thinker_idx:
          # Start with YAML defaults
          params = default_params.clone()

          # Override with user-provided values
          for field in msgspec.structs.fields(type(params)):
              if hasattr(request, field.name):
                  value = getattr(request, field.name)
                  if value is not None:
                      setattr(params, field.name, value)

          sampling_params_list.append(params)
```

Comment (tzhouam) [reply vllm_omni/entrypoints/openai/serving_chat.py]: I understand that you’re loading the default sampling parameters and overriding them with the passed-in max_tokens, which is great.

As mentioned in my previous comment, I’d like to see some tests for other sampling parameters as well, such as temperature, since this implementation seems work for them also, without any additional change as the to_samping_params() will load them to the sampling parameters.

If you prefer to keep this PR focused on max_tokens only, I’m also fine with leaving the other sampling parameters to the future work.

**Comment by wonjerry** [comment]
> This looks great! One more small suggestion, besides the naming problem: could you add some tests for parameters other than max_tokens (e.g. temperature)? The code seems correct from reading it, but a few extra tests would make it even more robust. Thanks a lot for your effort and contribution.

@tzhouam Thanks! I added test cases

**Comment by wonjerry** [comment]
@hsliuustc0106 @tzhouam please check
- Preserve YAML-defined custom sampling params (top_k, repetition_penalty, etc.) by using clone+override instead of request.to_sampling_params() which ignores them.
  - https://github.com/vllm-project/vllm-omni/pull/450/changes/c1cda7ed1277f92c54bbba75635fc0b430562073
- fix openai entrypoint test directory name becase of shadowing the real openai package during pytest execution
  - https://github.com/vllm-project/vllm-omni/pull/450/changes/a13f27756015ddcf0a80475fc1062731c3f9a462
  -  run successfully in my local

**Comment by tzhouam** [comment]
Nice.
Please also update the function name as above. The name "thinker" is too model‑specific and shouldn’t be used in the API server, which is designed for all models.

**Comment by wonjerry** [comment]
@tzhouam I agree. I rename thinker to comprehension. PTAL

**Review by tzhouam — APPROVED** [review]
LGTM

**Comment by wonjerry** [comment]
@tzhouam Thank you for the feedback. I'll keep this PR focused on max_tokens and add the additional test cases for other parameters in a future PR!

**Comment by hsliuustc0106** [comment]
thanks for your contribution, merged
